apps/sync/js/sync-engine/syncengine.js

```diff
@@ -0,0 +1,274 @@
+      var collection = this._getCollection(collectionName, isControlCollection);
+      // Let synchronization strategy default to 'manual', see
+      // http://kintojs.readthedocs.org \
+      //     /en/latest/api/#fetching-and-publishing-changes
+
+      return collection.sync()
+      .catch(err => {
+        throw err;
+      }).then(syncResults => {
+        if (syncResults.ok) {
+          return syncResults;
+        } else {
+          return Promise.reject(new SyncEngine.UnrecoverableError());
+        }
+      }).then(syncResults => {
+        if (syncResults.conflicts.length) {
+          return this._resolveConflicts(collectionName, syncResults.conflicts);
+        }
+      }).catch(err => {
+        if (err instanceof TypeError) {
+          throw new SyncEngine.UnrecoverableError();
+        } else if (err instanceof Error && typeof err.request === 'object') {
+          if (err.request.status === 401) {
+            throw new SyncEngine.AuthError();
+          } else {
+            throw new SyncEngine.TryLaterError();
+          }
+        } else {
+          throw new SyncEngine.UnrecoverableError();
+        }
+      });
+    },
+
+    _storageVersionOK: function(metaGlobal) {
+      var payloadObj;
+      try {
+        payloadObj = JSON.parse(metaGlobal.data.payload);
+      } catch(e) {
+        return false;
+      }
+      return (typeof payloadObj === 'object' &&
+          payloadObj.storageVersion === 5);
+    },
+
+    _initFxSyncWebCrypto: function(cryptoKeys) {
+      this._fswc = new FxSyncWebCrypto();
+      return this._fswc.setKeys(this._kB, cryptoKeys).then(() => {}, (err) => {
+        if (err === 'SyncKeys hmac could not be verified with current main ' +
+            'key') {
+          throw new SyncEngine.UnrecoverableError();
+        } else {
+          throw err;
+        }
+      });
+    },
+
+    connect: function() {
+      return this._syncCollection('meta', true)
+      .then(() => this._getItem('meta', 'global', true))
+      .then(metaGlobal => {
+        if (!this._storageVersionOK(metaGlobal)) {
+          return Promise.reject(new SyncEngine.UnrecoverableError(`Incompatible\
+ storage version or storage version not recognized.`));
+        }
+        return this._syncCollection('crypto', true);
+      }).then(() => this._getItem('crypto', 'keys', true))
+      .then((cryptoKeysRecord) => {
+        var cryptoKeys;
+        try {
+          cryptoKeys = JSON.parse(cryptoKeysRecord.data.payload);
+        } catch (e) {
+          return Promise.reject(new SyncEngine.UnrecoverableError(`Could not pa\
+rse crypto/keys payload as JSON`));
+        }
+        return cryptoKeys;
+      }).then((cryptoKeys) => {
+        //Cannot do this from the constructor because we need cryptoKeys first,
+        //and it's not nice to return a promise from a constructor:
+        return this._initFxSyncWebCrypto(cryptoKeys);
+      });
+    },
+
+    registerAdapter: function(collectionName, adapter) {
```

Comment:
What's the use case for registering an adapter dynamically? We can leave the option to do it this way, but it would be great if we could allow adapters registration via SyncEngine constructor.

Allowing adapters to be registered as constructor args and queuing sync requests until the crypto keys are retrieved, we could have an API as simple as:

    let syncEngine = new SyncEngine(options);
    syncEngine.syncNow(collections);